### PR TITLE
use taipy-assets favicon

### LINKED ISSUE #1244: BUG-Unable to change the Icon post launch
**Description**
Unable to change the icon that appears on the browser
image

Also, any way to add headers & header validations for the GETing of the html page for security purposes?

### LINKED ISSUE #1327: Analytics: Use remote default favicon
## Description

Gui can use an image hosted on a Taipy server for the default favicon. Each time a GUI application is launched, the browser tries to fetch this image. If no favicon is found on the remote, whatever the reason, if possible, use a local favicon provided with the package so we can still run the app without internet access if needed.

The icon could be stored in Git Hub, which would make it transparent for users and allow Git Hub to process the information for use directly.
## Motivation

The purpose is to get Taipy usage information with metadata for better analytics. The analytics part will be done in another ticket.

EDITED:
Link to the favicon: https://github.com/Avaiga/taipy-assets/blob/develop/favicon.png

EDITED: Add some documentation to explain how to change the favicon and use a local one for unconnected users.

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
resolves #1327
allow to change favicon dynamically
resolves #1244

`gui.set_favicon("https://raw.githubusercontent.com/Avaiga/taipy-assets/develop/favicon-3.1.png")`
`state.set_favicon("https://raw.githubusercontent.com/Avaiga/taipy-assets/develop/favicon-3.1.png")`
`gui.set_favicon("https://raw.githubusercontent.com/Avaiga/taipy-assets/develop/favicon-3.1.png", state)`

### CODE REVIEW

File: taipy/gui/gui.py [hunk 1]
```diff
@@ -223,6 +223,7 @@ class Gui:
     __ROBOTO_FONT = "https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap"
     __DOWNLOAD_ACTION = "__Taipy__download_csv"
     __DOWNLOAD_DELETE_ACTION = "__Taipy__download_delete_csv"
+    __FAVICON_URL = "https://raw.githubusercontent.com/Avaiga/taipy-assets/develop/favicon.png"
```
Comment (FabienLelaquais) [inline taipy/gui/gui.py]: I... guess I would rename this variable to '__DEFAULT_FAVICON_URL'? Makes it really long, I know...

File: taipy/gui/gui.py [hunk 2]
```diff
@@ -2621,3 +2637,16 @@ def stop(self):
 
     def _get_autorization(self, client_id: t.Optional[str] = None, system: t.Optional[bool] = False):
         return contextlib.nullcontext()
+
+    def set_favicon(self, favicon_path: t.Union[str, Path]):
+        """
```
Comment (FabienLelaquais) [inline taipy/gui/gui.py]: Same line, please. Just like the others
-> 
`"""Change the favicon for all clients.`

File: taipy/gui/gui.py [hunk 3]
```diff
@@ -2621,3 +2637,16 @@ def stop(self):
...
+        Change the favicon for all clients.
+
+        This function Allows to change dynamically the favicon for all connected client.
```
Comment (FabienLelaquais) [inline taipy/gui/gui.py]:
```suggestion
        This function dynamically changes the favicon of Taipy GUI pages for all connected client.
```

File: taipy/gui/gui.py [hunk 4]
```diff
@@ -2621,3 +2637,16 @@ def stop(self):
...
+        favicon_path can be an URL (relative or not) or a file path.
+        The *favicon* parameter to the `(Gui.)run^` can also be used to change the favicon at start.
```
Comment (FabienLelaquais) [inline taipy/gui/gui.py]:
```suggestion
        The *favicon* parameter to `(Gui.)run()^` can also be used to change the favicon when the application starts.
```

**Review by FabienLelaquais — DISMISSED** [review]
I love it!

File: taipy/gui/state.py [hunk 5]
```diff
@@ -243,3 +245,12 @@ def __enter__(self):
 
     def __exit__(self, exc_type, exc_value, traceback):
         return super().__getattribute__(State.__attrs[0]).__exit__(exc_type, exc_value, traceback)
+
+    def set_favicon(self, favicon_path: t.Union[str, Path]):
+        """Change the favicon for a specific client.
```
Comment (FabienLelaquais) [inline taipy/gui/state.py]:
```suggestion
        """Change the favicon for the client of this state.
```

File: taipy/gui/state.py [hunk 6]
```diff
@@ -243,3 +245,12 @@ def __enter__(self):
...
+
+        This function dynamically changes the favicon of Taipy GUI pages for a specific client.
```
Comment (FabienLelaquais) [inline taipy/gui/state.py:252]:
```suggestion
        This function dynamically changes the favicon of Taipy GUI pages for this client.
```

File: taipy/gui/gui.py [hunk 7]
```diff
@@ -2635,14 +2635,15 @@ def stop(self):
     def _get_autorization(self, client_id: t.Optional[str] = None, system: t.Optional[bool] = False):
         return contextlib.nullcontext()
 
-    def set_favicon(self, favicon_path: t.Union[str, Path]):
+    def set_favicon(self, favicon_path: t.Union[str, Path], state: t.Optional[State] = None):
         """Change the favicon for all clients.
 
         This function dynamically changes the favicon of Taipy GUI pages for all connected client.
         favicon_path can be an URL (relative or not) or a file path.
         The *favicon* parameter to `(Gui.)run()^` can also be used to change the favicon when the application starts.
```
Comment (FabienLelaquais) [inline taipy/gui/gui.py]:
```suggestion
        TODO The *favicon_url* parameter to `(Gui.)run()^` can also be used to change the favicon when the application starts.
```
(doc is kind of incomplete - but will do in the future)

**Review by FabienLelaquais — DISMISSED** [review]
👍

File: taipy/gui/gui.py (hunk 7 above)
Comment (FredLL-Avaiga) [reply taipy/gui/gui.py]: the run parameter is named `favicon`

**Review by FabienLelaquais — DISMISSED** [review]
👍

**Review by FabienLelaquais — APPROVED** [review]
🤞

**Pull request merged**